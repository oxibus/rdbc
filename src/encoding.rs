// https://github.com/hsivonen/recode_rs

use std::io::{Read, Write};
use std::str::from_utf8_mut;

use encoding_rs::*;

use crate::error::DbcError;

pub fn utf8_to_gbk(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "UTF-8", "GBK")
}

pub fn gbk_to_utf8(src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, "GBK", "UTF-8")
}

pub fn to_utf8(src_encoding_label: &str, src_data: &[u8]) -> Result<Vec<u8>, DbcError> {
    recode(src_data, src_encoding_label, "UTF-8")
}

pub fn recode(
    src_data: &[u8],
    src_encoding_label: &str,
    dst_encoding_label: &str,
) -> Result<Vec<u8>, DbcError> {
    let src_encoding = get_encoding(Some(src_encoding_label.to_string()))?;
    let dst_encoding = get_encoding(Some(dst_encoding_label.to_string()))?;

    let mut decoder = src_encoding.new_decoder();
    let mut encoder = dst_encoding.new_encoder();

    let mut buf = std::io::Cursor::new(Vec::new());

    convert_via_utf8(
        &mut decoder,
        &mut encoder,
        &mut std::io::Cursor::new(src_data),
        &mut buf,
        true,
    )?;

    Ok(buf.into_inner())
}

pub fn get_encoding(opt: Option<String>) -> Result<&'static Encoding, DbcError> {
    match opt {
        None => Ok(UTF_8),
        Some(label) => match Encoding::for_label(label.as_bytes()) {
            None => Err(DbcError::InvalidEncodingLabel(label)),
            Some(encoding) => Ok(encoding),
        },
    }
}

pub fn convert_via_utf8(
    decoder: &mut Decoder,
    encoder: &mut Encoder,
    read: &mut dyn Read,
    write: &mut dyn Write,
    last: bool,
) -> Result<(), DbcError> {
    let mut input_buffer = [0u8; 2048];
    let mut intermediate_buffer_bytes = [0u8; 4096];
    let intermediate_buffer: &mut str = from_utf8_mut(&mut intermediate_buffer_bytes[..]).unwrap();
    let mut output_buffer = [0u8; 4096];
    let mut current_input_ended = false;
    while !current_input_ended {
        match read.read(&mut input_buffer) {
            Err(e) => {
                log::error!("Error reading input, error = {}", e);
                return Err(DbcError::EncodingReadInputError);
            }
            Ok(decoder_input_end) => {
                current_input_ended = decoder_input_end == 0;
                let input_ended = last && current_input_ended;
                let mut decoder_input_start = 0usize;
                loop {
                    let (decoder_result, decoder_read, decoder_written, _) = decoder.decode_to_str(
                        &input_buffer[decoder_input_start..decoder_input_end],
                        intermediate_buffer,
                        input_ended,
                    );
                    decoder_input_start += decoder_read;

                    let last_output = if input_ended {
                        match decoder_result {
                            CoderResult::InputEmpty => true,
                            CoderResult::OutputFull => false,
                        }
                    } else {
                        false
                    };

                    // Regardless of whether the intermediate buffer got full
                    // or the input buffer was exhausted, let's process what's
                    // in the intermediate buffer.

                    if encoder.encoding() == UTF_8 {
                        // If the target is UTF-8, optimize out the encoder.
                        if let Err(e) =
                            write.write_all(&intermediate_buffer.as_bytes()[..decoder_written])
                        {
                            log::error!("Error writing output, error = {}", e);
                            return Err(DbcError::EncodingWriteOutputError);
                        }
                    } else {
                        let mut encoder_input_start = 0usize;
                        loop {
                            let (encoder_result, encoder_read, encoder_written, _) = encoder
                                .encode_from_utf8(
                                    &intermediate_buffer[encoder_input_start..decoder_written],
                                    &mut output_buffer,
                                    last_output,
                                );
                            encoder_input_start += encoder_read;
                            if let Err(e) = write.write_all(&output_buffer[..encoder_written]) {
                                log::error!("Error writing output, error = {}", e);
                                return Err(DbcError::EncodingWriteOutputError);
                            }
                            match encoder_result {
                                CoderResult::InputEmpty => {
                                    break;
                                }
                                CoderResult::OutputFull => {
                                    continue;
                                }
                            }
                        }
                    }

                    // Now let's see if we should read again or process the
                    // rest of the current input buffer.
                    match decoder_result {
                        CoderResult::InputEmpty => {
                            break;
                        }
                        CoderResult::OutputFull => {
                            continue;
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
