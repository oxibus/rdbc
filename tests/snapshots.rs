#![cfg(feature = "serde")]

use std::fs;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot, with_settings};
use rrdbc::ast::network_ast::parse_dbc;
use rrdbc::encoding::decode_cp1252;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [
        ("dbc/dbc_from_cantools", "dbc-cantools"),
        ("dbc/mytest", "mytest"),
    ];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(&path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) {
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match parse_dbc(buffer.as_ref()) {
            Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
            Err(e) => {
                eprintln!("Failed to parse {file_name}.dbc: {e:?}");
                assert_debug_snapshot!(format!("!error___{file_name}"), e);
            }
        }
    } else {
        let error = format!("Failed to decode {file_name}.dbc as cp1252");
        eprintln!("{error}");
        assert_snapshot!(format!("!error___{file_name}"), error);
    }
}
