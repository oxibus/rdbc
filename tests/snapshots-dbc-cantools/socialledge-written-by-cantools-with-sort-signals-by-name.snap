---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing:
  value: ~
nodes:
  - DBG
  - DRIVER
  - IO
  - MOTOR
  - SENSOR
value_tables: ~
messages:
  - header:
      id: 100
      name: DRIVER_HEARTBEAT
      size: 1
      transmitter: DRIVER
    signals:
      - name: DRIVER_HEARTBEAT_cmd
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - SENSOR
          - MOTOR
  - header:
      id: 500
      name: IO_DEBUG
      size: 4
      transmitter: IO
    signals:
      - name: IO_DEBUG_test_enum
        multiplexer: ~
        start_bit: 8
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: IO_DEBUG_test_float
        multiplexer: ~
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.5
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: IO_DEBUG_test_signed
        multiplexer: ~
        start_bit: 16
        size: 8
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: IO_DEBUG_test_unsigned
        multiplexer: ~
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
  - header:
      id: 101
      name: MOTOR_CMD
      size: 1
      transmitter: DRIVER
    signals:
      - name: MOTOR_CMD_drive
        multiplexer: ~
        start_bit: 4
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 9
        unit: ""
        receivers:
          - MOTOR
      - name: MOTOR_CMD_steer
        multiplexer: ~
        start_bit: 0
        size: 4
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: -5
        min: -5
        max: 5
        unit: ""
        receivers:
          - MOTOR
  - header:
      id: 400
      name: MOTOR_STATUS
      size: 3
      transmitter: MOTOR
    signals:
      - name: MOTOR_STATUS_speed_kph
        multiplexer: ~
        start_bit: 8
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.001
        offset: 0
        min: 0
        max: 0
        unit: kph
        receivers:
          - DRIVER
          - IO
      - name: MOTOR_STATUS_wheel_error
        multiplexer: ~
        start_bit: 0
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
  - header:
      id: 200
      name: SENSOR_SONARS
      size: 8
      transmitter: SENSOR
    signals:
      - name: SENSOR_SONARS_err_count
        multiplexer: ~
        start_bit: 4
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
      - name: SENSOR_SONARS_left
        multiplexer:
          multiplexer_signal: 0
          multiplexer_switch: ~
        start_bit: 16
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
      - name: SENSOR_SONARS_middle
        multiplexer:
          multiplexer_signal: 0
          multiplexer_switch: ~
        start_bit: 28
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
      - name: SENSOR_SONARS_mux
        multiplexer:
          multiplexer_signal: ~
          multiplexer_switch: ~
        start_bit: 0
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
      - name: SENSOR_SONARS_no_filt_left
        multiplexer:
          multiplexer_signal: 1
          multiplexer_switch: ~
        start_bit: 16
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: SENSOR_SONARS_no_filt_middle
        multiplexer:
          multiplexer_signal: 1
          multiplexer_switch: ~
        start_bit: 28
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: SENSOR_SONARS_no_filt_rear
        multiplexer:
          multiplexer_signal: 1
          multiplexer_switch: ~
        start_bit: 52
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: SENSOR_SONARS_no_filt_right
        multiplexer:
          multiplexer_signal: 1
          multiplexer_switch: ~
        start_bit: 40
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DBG
      - name: SENSOR_SONARS_rear
        multiplexer:
          multiplexer_signal: 0
          multiplexer_switch: ~
        start_bit: 52
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
      - name: SENSOR_SONARS_right
        multiplexer:
          multiplexer_signal: 0
          multiplexer_switch: ~
        start_bit: 40
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - DRIVER
          - IO
env_vars: []
env_vars_data: []
comments:
  - Node:
      node_name: DRIVER
      comment: // The driver controller driving the car //
  - Node:
      node_name: MOTOR
      comment: The motor controller of the car
  - Node:
      node_name: SENSOR
      comment: The sensor controller of the car
  - Message:
      message_id: 100
      comment: Sync message used to synchronize the controllers
attribute_definitions:
  - Network:
      attribute_name: BusType
      attribute_value_type:
        String: {}
  - Message:
      attribute_name: GenMsgCycleTime
      attribute_value_type:
        Integer:
          minimum: 0
          maximum: 0
  - Signal:
      attribute_name: FieldType
      attribute_value_type:
        String: {}
attribute_defaults:
  - Attribute:
      attribute_name: BusType
      attribute_value:
        String: CAN
  - Attribute:
      attribute_name: GenMsgCycleTime
      attribute_value:
        Double: 0
  - Attribute:
      attribute_name: FieldType
      attribute_value:
        String: ""
attribute_values:
  - Message:
      attribute_name: GenMsgCycleTime
      message_id: 100
      attribute_value:
        Double: 1000
  - Signal:
      attribute_name: FieldType
      message_id: 100
      signal_name: DRIVER_HEARTBEAT_cmd
      attribute_value:
        String: DRIVER_HEARTBEAT_cmd
  - Message:
      attribute_name: GenMsgCycleTime
      message_id: 500
      attribute_value:
        Double: 100
  - Signal:
      attribute_name: FieldType
      message_id: 500
      signal_name: IO_DEBUG_test_enum
      attribute_value:
        String: IO_DEBUG_test_enum
  - Message:
      attribute_name: GenMsgCycleTime
      message_id: 101
      attribute_value:
        Double: 100
  - Message:
      attribute_name: GenMsgCycleTime
      message_id: 400
      attribute_value:
        Double: 100
  - Message:
      attribute_name: GenMsgCycleTime
      message_id: 200
      attribute_value:
        Double: 100
signal_value_descriptions:
  - message_id: 100
    signal_name: DRIVER_HEARTBEAT_cmd
    value_descriptions:
      values:
        - num: 2
          str: DRIVER_HEARTBEAT_cmd_REBOOT
        - num: 1
          str: DRIVER_HEARTBEAT_cmd_SYNC
        - num: 0
          str: DRIVER_HEARTBEAT_cmd_NOOP
  - message_id: 500
    signal_name: IO_DEBUG_test_enum
    value_descriptions:
      values:
        - num: 2
          str: IO_DEBUG_test2_enum_two
        - num: 1
          str: IO_DEBUG_test2_enum_one
env_var_value_descriptions: []
